internal/controller/apiextensions/composite/reconciler.go

```diff
@@ -399,6 +471,17 @@ func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error)
 			return reconcile.Result{RequeueAfter: shortWait}, nil
 		}
 
+		// DeepCopy composed resource to ensure changes to composed are not persisted
+		// when fetching connection details or readiness
+		cdCopy := cd.DeepCopyObject()
+		for _, p := range comp.Spec.Resources[i].Patches {
+			if err := p.Apply(patchObject, cdCopy); err != nil {
+				log.Debug(fmt.Sprintf(errFmtPatch, i), "error", err)
+				r.record.Event(cr, event.Warning(reasonCompose, err))
+				return reconcile.Result{RequeueAfter: shortWait}, nil
+			}
+		}
```

Comment:
What do you think about:

1. Moving this patch loop into a new `Renderer`
1. Adding an optional patch type filter to `p.Apply`.

This would keep all of the patch application at the same 'scope' (e.g. in methods called by the reconciler, rather than partly in the reconciler and partly in said methods) and allow us to avoid spending time calculating patches that we don't need.

I'm thinking something like:

```go
// Apply executes a patching operation between the from and to resources.
// If applies all patch types unless an 'only' filter is supplied.
func (c *Patch) Apply(from, to runtime.Object, only ...PatchType) error {
	// ...
}
```

With the reconciler being something like:

```go
// Render each composed resource - applies only the composite -> composed patches.
for i := range refs {
	cd := composed.New(composed.FromReference(refs[i]))
	_ = r.composed.Render(ctx, cr, cd, comp.Spec.Resources[i])
}

// Persist references
cr.SetResourceReferences(refs)
_ := r.client.Update(ctx, cr)

// Similar to the current loop, but we also 're-render' the composite resource based on each
// composed resource.
for i, cd := range cds {
	_ = r.client.Apply(ctx, cd, resource.MustBeControllableBy(cr.GetUID()))
	c, _ := r.composed.FetchConnectionDetails(ctx, cd, comp.Spec.Resources[i])
	rdy, _ := r.composed.IsReady(ctx, cd, comp.Spec.Resources[i])

	// Note composite.Render, rather than composed.Render. Applies only the
	// composed -> composite patches.
	_ = r.composite.Render(ctx, cr, cd, comp.Spec.Resources[i])

}

```

Reply:
A composite render is a good idea, I tried to re-use the composite render initially but that path seemed unsuccessful 😄